docs/basics/101-129-python.rst

```diff
@@ -0,0 +1,79 @@
+
+DataLad's Python API
+--------------------
+
+.. note::
+
+   While there is a dedicated API for Python, DataLad's functions can of course
+   also be used with other programming languages, such as Matlab, via standard
+   system calls.
+
+"Whatever you can do with DataLad from the command line, you can also do it with
+DataLads Python API", begins the lecturer.
+"In addition to the command line interface you are already very familiar with,
+DataLad's functionality can also be used within interactive Python sessions
+or Python scripts.
+This feature can help to automate dataset operations, provides an alternative
+to the command line, and it is immensely useful when creating reproducible
+data analyses."
+
+Todays lecture will give you an overview on DataLad's Python API and explore
+how to make use of it in an analysis project. Together with the previous
+section on the YODA principles, this section will be the basis for a data
+analysis midterm project.
+
+All of DataLad's user-oriented commands are exposed via ``datalad.api``.
+Thus, any command can be imported as a stand-alone command like this::
+
+   >>> from datalad.api import <COMMAND>
+
+The `developer documentation <http://docs.datalad.org/en/latest/modref.html>`_
+of DataLad lists an overview of all commands, but naming is congruent to the
+command line interface. The only functionality that is not available at the
+command line is ``datalad.api.Dataset``, DataLads core data type
+representation in Python. Just like any other command, it can be imported
+like this::
+
+   >>> from datalad.api import Dataset
+
+
+A ``Dataset`` is a class that represents a DataLad dataset. In addition to the
+stand-alone commands, all of DataLad's functionality is available via *methods*
+of this class as well. Thus, these are two equally valid ways to create a new
+dataset with DataLad in Python::
+
+   >>> from datalad.api import create, Dataset
+   # create as a stand-alone command
+   >>> create(path='scratch/test')
+   [INFO   ] Creating a new annex repo at /home/me/scratch/test
+   Out[3]: <Dataset path=/home/me/scratch/test>
+
+   # create as a dataset method
+   >>> ds = Dataset(path='scratch/test')
+   >>> ds.create()
+   [INFO   ] Creating a new annex repo at /home/me/scratch/test
+   Out[3]: <Dataset path=/home/me/scratch/test>
+
+As shown above, the only required parameter for a Dataset is the ``path`` to
+its location, and this location may or may not exist yet.
+
+Use cases for DataLad's Python API
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+"Why should one use the Python API? Can we not do everything necessary via the
+command line already? Does Python add anything to this?" asks somebody.
+
+It is completely up to on you and dependent on your preferred workflow
+whether you decide to use the command line or the Python API of DataLad for
+the majority of tasks. Both are valid ways to accomplish the same results.
+One typical application though is the use of :command:`datalad get` in Python
```

Comment:
Everything that follows is more an example of what can "also be done in Python". I see the danger that it comes across as "better be done in Python", and consequently communicates that Python is the "favorite child" -- which it really isn't.

There is one major advantage of using the Python API though, and it is exactly the one you start with: the Dataset. Via the command line a dataset needs to be discovered and inspected with each datalad command call, because there is no persistence of information in a, e.g., shell script. Given that the `datalad` command has a startup time (even when doing nothing) of ~200ms, this means that there is the potential for substantial speed-up when doing many calls to the API, and using a persistent `Dataset` object instance.

It might be better to use this as an example, because a single call to `get` should be as fast via both APIs.

Reply:
Thanks for that! I'll change this. Thinking about this further, I think I will try to "downgrade" this section on Python to a subsection or hidden section. This way, the whole chapter will have a major focus on YODA (both in theory and practice), with some added benefit on learning about the Python API. This may also resolve the chaotic summary-situation (multiple summaries in one chapter) here, might be more intuitively structured, and shifts the focus from the Python API. No guarantee,  I'll see whether I can make this work.